### PR TITLE
add support for zipfiles with implicit directories

### LINKED ISSUE #36: finder fails for CDSE SAFE ZIP files without implicit directoruies
Since Sep 26th, files packaged with Copernicus Data Service Ecosystem do not include implicit directories in the archive namelist.

These are valid zipfiles, but this leads to issues with packages like PyroSAR that use this to indicate if an archive is.

Example namelists

Old SAFE archive _**with**_ directories in namelist:
```
['S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/',
 'S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE-report-20231005T214402.pdf',
 'S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/annotation/',
 'S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/annotation/calibration/',
 'S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/annotation/calibration/calibration-s1a-iw-grd-vh-20231005t190129-20231005t190154-050634-06199a-002.xml',
 'S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/annotation/calibration/calibration-s1a-iw-grd-vv-20231005t190129-20231005t190154-050634-06199a-001.xml',
 'S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/annotation/calibration/noise-s1a-iw-grd-vh-20231005t190129-20231005t190154-050634-06199a-002.xml',
 'S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/annotation/calibration/noise-s1a-iw-grd-vv-20231005t190129-20231005t190154-050634-06199a-001.xml',
 'S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/annotation/rfi/',
 'S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/annotation/rfi/rfi-s1a-iw-grd-vh-20231005t190129-20231005t190154-050634-06199a-002.xml',
 'S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/annotation/rfi/rfi-s1a-iw-grd-vv-20231005t190129-20231005t190154-050634-06199a-001.xml',
 'S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/annotation/s1a-iw-grd-vh-20231005t190129-20231005t190154-050634-06199a-002.xml',
 'S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/annotation/s1a-iw-grd-vv-20231005t190129-20231005t190154-050634-06199a-001.xml',
 'S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/manifest.safe',
 'S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/measurement/',
 'S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/measurement/s1a-iw-grd-vh-20231005t190129-20231005t190154-050634-06199a-002.tiff',
 'S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/measurement/s1a-iw-grd-vv-20231005t190129-20231005t190154-050634-06199a-001.tiff',
 'S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/preview/',
 'S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/preview/icons/',
 'S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/preview/icons/logo.png',
 'S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/preview/map-overlay.kml',
 'S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/preview/product-preview.html',
 'S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/preview/quick-look.png',
 'S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/support/',
 'S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/support/s1-level-1-calibration.xsd',
 'S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/support/s1-level-1-measurement.xsd',
 'S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/support/s1-level-1-noise.xsd',
 'S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/support/s1-level-1-product.xsd',
 'S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/support/s1-level-1-quicklook.xsd',
 'S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/support/s1-level-1-rfi.xsd',
 'S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/support/s1-map-overlay.xsd',
 'S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/support/s1-object-types.xsd',
 'S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/support/s1-product-preview.xsd']
```

SAFE archive **_without_** implicit directories in namelist:
```
['S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE-report-20231005T214402.pdf',
 'S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/annotation/calibration/calibration-s1a-iw-grd-vh-20231005t190129-20231005t190154-050634-06199a-002.xml',
 'S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/annotation/calibration/calibration-s1a-iw-grd-vv-20231005t190129-20231005t190154-050634-06199a-001.xml',
 'S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/annotation/calibration/noise-s1a-iw-grd-vh-20231005t190129-20231005t190154-050634-06199a-002.xml',
 'S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/annotation/calibration/noise-s1a-iw-grd-vv-20231005t190129-20231005t190154-050634-06199a-001.xml',
 'S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/annotation/rfi/rfi-s1a-iw-grd-vh-20231005t190129-20231005t190154-050634-06199a-002.xml',
 'S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/annotation/rfi/rfi-s1a-iw-grd-vv-20231005t190129-20231005t190154-050634-06199a-001.xml',
 'S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/annotation/s1a-iw-grd-vh-20231005t190129-20231005t190154-050634-06199a-002.xml',
 'S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/annotation/s1a-iw-grd-vv-20231005t190129-20231005t190154-050634-06199a-001.xml',
 'S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/manifest.safe',
 'S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/measurement/s1a-iw-grd-vh-20231005t190129-20231005t190154-050634-06199a-002.tiff',
 'S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/measurement/s1a-iw-grd-vv-20231005t190129-20231005t190154-050634-06199a-001.tiff',
 'S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/preview/icons/logo.png',
 'S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/preview/map-overlay.kml',
 'S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/preview/product-preview.html',
 'S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/preview/quick-look.png',
 'S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/preview/thumbnail.png',
 'S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/support/s1-level-1-calibration.xsd',
 'S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/support/s1-level-1-measurement.xsd',
 'S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/support/s1-level-1-noise.xsd',
 'S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/support/s1-level-1-product.xsd',
 'S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/support/s1-level-1-quicklook.xsd',
 'S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/support/s1-level-1-rfi.xsd',
 'S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/support/s1-map-overlay.xsd',
 'S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/support/s1-object-types.xsd',
 'S1A_IW_GRDH_1SDV_20231005T190129_20231005T190154_050634_06199A_4144.SAFE/support/s1-product-preview.xsd']
 ```

### PR DESCRIPTION
**Pull request by tyler-c2s**
closes #36 

Iterates through namelist of archive, and retrieves the parent directories of each file as a set

Created test zipfile with `zip -r -D ../demo_finder_noimplicit.zip * `

### CODE REVIEW

File: spatialist/ancillary.py [hunk 1]
```diff
@@ -23,6 +23,8 @@
 import subprocess as sp
 import tarfile as tf
 import zipfile as zf
+from typing import Iterable, List
+from zipfile import ZipFile
```
Comment (johntruckenbrodt) [inline spatialist/ancillary.py:27]: `zipfile` is already imported above. Can you please combine it into one?

File: spatialist/ancillary.py [hunk 2]
```diff
@@ -114,6 +116,47 @@ def dissolve(inlist):
         out.extend(dissolve(i)) if isinstance(i, list) else out.append(i)
     return out
 
+def parent_dirs(path: str) -> Iterable[str]:
+    """
+    generator that yeilds parent directories of a zipfile path
```
Comment (johntruckenbrodt) [inline spatialist/ancillary.py:121]: Little typo:
```diff
- yeilds
+ yields
```

**Review by johntruckenbrodt — CHANGES_REQUESTED** [review]
Hi @tyler-c2s,
thanks a lot for implementing this. Here just a few minor corrections. Furthermore, do you know in which Python version this syntax was introduced:
```python
def namelist_with_implicit_dirs(root: ZipFile) -> List[str]:
```
The `pyproject.toml` currently specifies `requires-python = ">=3.0"`, which should be updated accordingly.

**Comment by tyler-c2s** [comment]
> Furthermore, do you know in which Python version this syntax was introduced:
> 
> ```python
> def namelist_with_implicit_dirs(root: ZipFile) -> List[str]:
> ```
> 
> The `pyproject.toml` currently specifies `requires-python = ">=3.0"`, which should be updated accordingly.

Great point. Typing was introduced in py 3.5. Happy to limit to that version, or as python <= 3.7 is now End Of Life we could limit the version to >=3.8

**Review by johntruckenbrodt — APPROVED** [review]

**Comment by johntruckenbrodt** [comment]
Great. Thanks. I have quickly changed it myself. Let's stay with 3.5 as minimum requirement.

**Comment by tyler-c2s** [comment]
appreciated, thanks for the quick review! :rocket:

**Pull request merged**